src/stream.rs

```diff
@@ -0,0 +1,301 @@
+// use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software is furnished to do so,
+// subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
+// FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS
+// OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,
+// WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
+// CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+use bytes::BytesMut;
+use error::StreamError;
+use frame::Body;
+use futures::{prelude::*, stream::{Fuse, Stream as FuturesStream}, sync::mpsc, task::{self, Task}};
+use std::{cmp::min, fmt, io, sync::{atomic::{AtomicUsize, Ordering}, Arc}, u32, usize};
+use tokio_io::{AsyncRead, AsyncWrite};
+use Config;
+
+
+#[derive(Copy, Clone, Debug, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct Id(u32);
+
+impl Id {
+    pub(crate) fn new(id: u32) -> Id {
+        Id(id)
+    }
+
+    pub fn is_server(&self) -> bool {
+        self.0 % 2 == 0
+    }
+
+    pub fn is_client(&self) -> bool {
+        !self.is_server()
+    }
+
+    pub fn is_session(&self) -> bool {
+        self.0 == 0
+    }
+
+    pub fn as_u32(&self) -> u32 {
+        self.0
+    }
+}
+
+impl fmt::Display for Id {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.0)
+    }
+}
+
+
+#[derive(Copy, Clone, Debug, PartialEq, Eq)]
+enum State {
+    Open,
+    SendClosed,
+    RecvClosed,
+    Closed
+}
+
+
+#[derive(Debug)]
+pub(crate) struct Window(AtomicUsize);
+
+impl Window {
+    pub(crate) fn new(n: AtomicUsize) -> Window {
+        Window(n)
+    }
+
+    pub(crate) fn decrement(&self, amount: usize) -> usize {
+        loop {
+            let prev = self.0.load(Ordering::SeqCst);
+            let next = prev.checked_sub(amount).unwrap_or(0);
+            if self.0.compare_and_swap(prev, next, Ordering::SeqCst) == prev {
+                return next
+            }
+        }
+    }
+
+    pub(crate) fn set(&self, val: usize) {
+        self.0.store(val, Ordering::SeqCst)
+    }
+}
+
+
+#[derive(Debug)]
+pub(crate) enum Item {
+    Data(Body),
+    WindowUpdate(u32),
+    Reset,
+    Finish
+}
+
+
+pub(crate) type Sender = mpsc::UnboundedSender<(Id, Item)>;
+pub(crate) type Receiver = Fuse<mpsc::UnboundedReceiver<Item>>;
+
+
+pub struct Stream {
+    id: Id,
+    state: State,
+    config: Arc<Config>,
+    recv_window: Arc<Window>,
+    send_window: u32,
+    buffer: BytesMut,
+    sender: Sender,
+    receiver: Receiver,
+    writer_task: Option<Task>
```

Comment:
Shouldn't that be a `Vec` (or something similar) in case multiple different tasks try to write "simultaneously"?